### PR TITLE
Feature/semester plan

### PR DESCRIPTION
**Pull request by cwngan**
#16

### CODE REVIEW

**Review by chemistrying — COMMENTED** [review]
I took a brief look and noticed the above issues, you may have a look

File: flaskr/api/semesterplan.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            SemesterPlanResponseModel(
+                status="ERROR", error="Unauthorized"
+            ).model_dump(),
+            401,
+        )
+    if not ObjectId.is_valid(semester_plan_id):
+        return (
+            SemesterPlanResponseModel(
+                status="ERROR", error="Malformed ID"
+            ).model_dump(),
```
[the remark this answers is not in the document]
Comment (lawjacob) [reply flaskr/api/semesterplan.py:128]: i think leaving a model_dump would enhcance the readability?

[the remark this answers is not in the document]
Comment (cwngan) [reply flaskr/api/semesterplan.py]: I don't think it helps readability? Also these kind of things should be consistent across different files.

File: flaskr/api/__init__.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 from flask import Blueprint
 from flask_pydantic import validate  # type: ignore
 
-from flaskr.api import health, ping, user, courses, course_plans
+from flaskr.api import health, ping, user, courses, course_plans,semester_plan
```
Comment (cwngan) [inline flaskr/api/__init__.py:4]: Please run `black .`

File: flaskr/api/semester_plan.py
Comment (cwngan) [inline flaskr/api/semester_plan.py:1]: Should be named `semester_plans`?

File: flaskr/db/semesterplan.py
Comment (cwngan) [inline flaskr/db/semesterplan.py:1]: Again, this should also be named `semester_plans.py`

**Review by cwngan — COMMENTED** [review]
Make sure to update all the related imports and format documents as well.

File: flaskr/api/__init__.py [hunk 3]
```diff
@@ -18,3 +18,8 @@ def root():
 route.register_blueprint(courses.route)
 route.register_blueprint(user.route)
 route.register_blueprint(course_plans.route)
+route.register_blueprint(semester_plans.route)
+
+
+# def register_blueprints(app):
```
Comment (chemistrying) [inline flaskr/api/__init__.py:24]: Wait... Why is there a commented code block... 👀

Comment (chemistrying) [inline flaskr/api/__init__.py:24]: Wait... Why is there a commented code block... 👀

File: flaskr/api/semester_plans.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            SemesterPlanResponseModel(
+                status="ERROR", error="Semester plan not found"
+            ).model_dump(),
+            404,
+        )
+    semester_plan_read = SemesterPlanRead.model_validate(semester_plan.model_dump())
+    semester_plan_read.created_at = semester_plan.created_at
+    return (
+        SemesterPlanResponseModel(
+            status="OK", data=semester_plan_read.model_dump()
```
Comment (chemistrying) [inline flaskr/api/semester_plans.py:67]: I don't think you need to add status="OK" in here because our response models default to status="OK"

File: flaskr/api/semester_plans.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    if not updated_plan:
+        return (
+            SemesterPlanResponseModel(
+                status="ERROR", error="Semester plan not found"
+            ).model_dump(),
+            404,
+        )
+    return (
+        SemesterPlanResponseModel(
+            status="OK", data=updated_plan.model_dump()
```
Comment (chemistrying) [inline flaskr/api/semester_plans.py:142]: Same as above

File: flaskr/api/semester_plans.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    if not deleted_plan:
+        return (
+            SemesterPlanResponseModel(
+                status="ERROR", error="Semester plan not found"
+            ).model_dump(),
+            404,
+        )
+    return (
+        SemesterPlanResponseModel(
+            status="OK", data=deleted_plan.model_dump()
```
Comment (chemistrying) [inline flaskr/api/semester_plans.py:176]: Same as above

File: flaskr/api/semester_plans.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+# Configure logging
+# logging.basicConfig(level=logging.DEBUG)
+# logger = logging.getLogger(__name__)
+
+route = Blueprint("semester_plans", __name__, url_prefix="/semester-plans")
+
+
+# Custom Exception
+class InvalidCredential(Exception):
```
Comment (chemistrying) [inline flaskr/api/semester_plans.py:29]: You haven't used this exception in this file...

File: flaskr/api/semester_plans.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+from flaskr.db.semester_plans import (
+    create_semester_plan,
+    get_semester_plan,
+    update_semester_plan,
+    delete_semester_plan,
+)
+from flaskr.db.user import get_user_by_username
+from flaskr.db.course_plans import get_course_plan  # Corrected the import
+
+# Configure logging
```
Comment (chemistrying) [inline flaskr/api/semester_plans.py:21]: Let's see how we can add a more convenient logging later... For now it's kind of annoying to perform logging in Flask servers... But for now let's remove this temporary code first

File: flaskr/db/database.py [hunk 9]
```diff
@@ -81,6 +81,9 @@ def init_db():
 
     db.users.create_index("email", unique=True)
     db.users.create_index("username", unique=True, sparse=True)
+    db.semester_plans.create_index("course_plan_id")
+    # db.semester_plans.create_index("semester")
```
Comment (chemistrying) [inline flaskr/db/database.py:85]: Don't think we need these comments...

File: flaskr/db/models.py [hunk 10]
```diff
@@ -73,6 +73,28 @@ class SemesterPlan(CoreModel):
...
     semester: int
     year: int
+    created_at: datetime
+
+
+class SemesterPlanRead(CoreModel):
+    id: Optional[PydanticObjectId] = Field(alias="_id", default=None)
+    course_plan_id: PydanticObjectId
+    courses: list[str]
+    semester: int
```
Comment (chemistrying) [inline flaskr/db/models.py:83]: I recommend the semester plan should be stored with `enum` data type. Then others will have an idea on what the semesters really are.

Comment (chemistrying) [reply flaskr/db/models.py:83]: Idk if there is `enum` there will be validation against the actual integer value (like can it be 4 or 400 or something else)

File: tests/test_api_semester_plans.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+
+@pytest.fixture
+def logged_in_client(test_user, client: FlaskClient):
+    with client.session_transaction() as session:
+        session["username"] = test_user.username
+    yield client
+
+
+def test_create_semester_plan(logged_in_client, test_course_plan):
```
Comment (chemistrying) [inline tests/test_api_semester_plans.py:36]: In general, you should make invalid testcases / boundary cases to check whether the endpoints are working properly / as intended

Comment (chemistrying) [reply tests/test_api_semester_plans.py:36]: You can make use of parametrization for more testcases, if you want to

**Review by chemistrying — CHANGES_REQUESTED** [review]
I just checked back the code and add more reviews

**Comment by chemistrying** [comment]
Now I'm just thinking that the whole integration should be done in this branch because it doesn't make any sense to separate the integration part - because for now we don't know whether these code will work...

**Comment by chemistrying** [comment]
I'll fix the stuff that doesn't affect the functionality and documentation a lot first... 👀

**Comment by chemistrying** [comment]
I added enforcement of semester and year compound key uniqueness since you shouldn't have two year 2025 and semester 1 semester plans, others are miscellaneous fixes (from what I've remembered), please review thanks 🙏

**Comment by FieryRMS** [comment]
Yall should have a look at #32 and merge with this branch if ok, i think will be useful here

**Comment by FieryRMS** [comment]
Handling the exact same logic individually for each function will make it very hard to test or change

**Review by FieryRMS — CHANGES_REQUESTED** [review]
Tests LGTM

File: flaskr/api/semester_plans.py
[the remark this answers is not in the document]
Comment (cwngan) [reply flaskr/api/semester_plans.py:1]: ~~I think I started this trend in course plans.~~ Anyways, yeah I think it should be more descriptive

File: flaskr/api/semester_plans.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+@validate(response_by_alias=True)
+def get(semester_plan_id):
+    """
+    Return the SemesterPlan with the specified id.
+    """
+    if not ObjectId.is_valid(semester_plan_id):
+        return (
+            SemesterPlanResponseModel(status="ERROR", error="Malformed ID"),
+            400,
+        )
```
[the remark this answers is not in the document]
Comment (cwngan) [reply flaskr/api/semester_plans.py:36]: Maybe then return 404 instead? IDK

[the remark this answers is not in the document]
Comment (FieryRMS) [reply flaskr/api/semester_plans.py:36]: yeah, same thoughts

File: flaskr/api/semester_plans.py
[the remark this answers is not in the document]
Comment (FieryRMS) [reply flaskr/api/semester_plans.py:1]: ~i didnt review that one so idk~

File: flaskr/api/semester_plans.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cwngan) [reply flaskr/api/semester_plans.py:36]: I think this also needs to be changed in course plan then 👀

[the remark this answers is not in the document]
Comment (FieryRMS) [reply flaskr/api/semester_plans.py:36]: just a side note: this can be solved as a side-effect of this [issue](https://github.com/cwngan/cu2m-backend/pull/20#discussion_r2071156318)

File: flaskr/api/semester_plans.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    semester_plan = get_semester_plan(semester_plan_id)
+    if not semester_plan:
+        return (
+            SemesterPlanResponseModel(status="ERROR", error="Semester plan not found"),
+            404,
+        )
+    return (
+        SemesterPlanResponseModel(data=semester_plan.model_dump()),
+        200,
+    )
```
[the remark this answers is not in the document]
Comment (chemistrying) [reply flaskr/api/semester_plans.py:46]: Yep, I'm shocked that I didn't notice that part... 💀

[the remark this answers is not in the document]
Comment (lawjacob) [reply flaskr/api/semester_plans.py:46]: we just need to after fetching semester plan, we also fetch course_plans' userid and check if it mataches authenticated user's id i think?

File: flaskr/api/semester_plans.py
[the remark this answers is not in the document]
Comment (lawjacob) [reply flaskr/api/semester_plans.py:1]: ~~idk i just follow d&i lol~~

File: flaskr/db/semester_plans.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    return SemesterPlan.model_validate(doc) if doc else None
+
+
+def delete_semester_plan(semester_plan_id: str):
+    db = get_db().semester_plans
+    doc = db.find_one_and_delete({"_id": ObjectId(semester_plan_id)})
+    return SemesterPlan.model_validate(doc) if doc else None
+
+
+def get_semester_plans_by_course_plan(course_plan_id: str):
```
Comment (cwngan) [inline flaskr/db/semester_plans.py:64]: Ok now IDK if `course_plan_id` should directly be `ObjectId` or not since the only place this is used currently is at getting course plan. That function converts the ObjectId to string and passes the string into this function, where it is converted into ObjectId again. But then it feels like we shouldn't change the type here just because other functions are accepting `str` 😔

File: flaskr/api/respmodels.py [hunk 15]
```diff
@@ -30,5 +30,15 @@ class UserResponseModel(ResponseModel):
...
+
+
 class CoursePlanResponseModel(ResponseModel):
     data: CoursePlanRead | list[CoursePlanRead] | None = None
+
+
+class CoursePlanWithSemestersResponseModel(ResponseModel):
+    data: dict | None = (
+        None  # {'course_plan': CoursePlanRead, 'semester_plans': list[SemesterPlanRead]}
+    )
```
Comment (cwngan) [inline flaskr/api/respmodels.py:44]: I think this can be more well-defined other than just a simple `dict` 💀

**Review by cwngan — COMMENTED** [review]
A quick review on the commits just now

File: flaskr/api/semester_plans.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (chemistrying) [reply flaskr/api/semester_plans.py]: Resolved in [671a5bb](https://github.com/cwngan/cu2m-backend/pull/20/commits/671a5bbb400e496894f283cce5acac26cea4d292)

File: flaskr/api/semester_plans.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    Create a SemesterPlan with the given parameters.
+    """
+    # Validate and convert course_plan_id
+    try:
+        course_plan_id = ObjectId(body.course_plan_id)
+    except Exception:
+        return (
+            SemesterPlanResponseModel(status="ERROR", error="Invalid course_plan_id"),
+            400,
+        )
```
[the remark this answers is not in the document]
Comment (chemistrying) [reply flaskr/api/semester_plans.py]: Resolved in [671a5bb](https://github.com/cwngan/cu2m-backend/pull/20/commits/671a5bbb400e496894f283cce5acac26cea4d292)

File: flaskr/api/semester_plans.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (chemistrying) [reply flaskr/api/semester_plans.py]: Resolved in [671a5bb](https://github.com/cwngan/cu2m-backend/pull/20/commits/671a5bbb400e496894f283cce5acac26cea4d292)

File: flaskr/api/semester_plans.py
[the remark this answers is not in the document]
Comment (chemistrying) [reply flaskr/api/semester_plans.py:1]: Resolved in [671a5bb](https://github.com/cwngan/cu2m-backend/pull/20/commits/671a5bbb400e496894f283cce5acac26cea4d292)

File: flaskr/api/respmodels.py (hunk 15 above)
Comment (chemistrying) [reply flaskr/api/respmodels.py:44]: Fr this is 💀

Comment (lawjacob) [reply flaskr/api/respmodels.py:44]: 💀

Comment (chemistrying) [reply flaskr/api/respmodels.py:44]: /j btw, I was a little bit tired lol

File: flaskr/db/semester_plans.py (hunk 14 above)
Comment (chemistrying) [reply flaskr/db/semester_plans.py:64]: It really depends, I think both is fine although I prefer it to be ObjectId, but then you gotta refactor to make every part consistent which is kind of annoying
If it works then just don't touch it I think unless we have time 👀

Comment (lawjacob) [reply flaskr/db/semester_plans.py:64]: if all other functions are accepting `str`, why don't we use `str` as the protocol between all the functions? that is more consistent i guess?

Comment (cwngan) [reply flaskr/db/semester_plans.py:64]: It is more consistent. I guess it's just me then because I just feel like that is doing unnecessary conversion twice. Anyways, I guess we can just keep it like this for now.

Comment (FieryRMS) [reply flaskr/db/semester_plans.py]: while resolving typing issues, i had to solve this. My target wasnt consistency, so some files might still be using str, some ObjectId. It should be consistent end-to-end, just not throughout whole repo. Marking as resolved by https://github.com/cwngan/cu2m-backend/pull/20/commits/0afa0bc4e54ccc98628ace06a3ae17b1bd19ee23

**Review by chemistrying — APPROVED** [review]
I think it's ok... Maybe...

**Review by FieryRMS — APPROVED** [review]
tests working, lgtm 👍

**Comment by chemistrying** [comment]
I realized that you can change the course list to arbitrary anything... Idk if that's intended or not, but if that's intended then you need to ensure that client side rendering won't break because of invalid course code 👀

**Comment by FieryRMS** [comment]
Wait a minute ☠️

**Comment by FieryRMS** [comment]
Yeah i think this should be upto the frontend to handle. Since frontend has to request for the courses anyway, it automatically filter itself, so backend doesnt have to repeat this logic just to update user course list

**Comment by cwngan** [comment]
I was thinking that backend should at least tell frontend that it has problem but I guess it's fair to leave to frontend to handle all of these. I think it should be just a try catch in an axios statement (I hope...)

**Comment by FieryRMS** [comment]
Actually if this is the way we handle it now, should we make course endpoint allow invalid course code and filter them? One way this can happen naturally when some courses get removed from database. So users may get softlocked (depending on implementation).

**Comment by FieryRMS** [comment]
Wait what am i babbling about, if delete course it still pass regex 🤦

**Comment by cwngan** [comment]
IT'S MERGED YES!

**Comment by FieryRMS** [comment]
😭
